// state-pool/esm/Store.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createStore = exports.Store = void 0;
var GlobalState_1 = require("./GlobalState");
var useGlobalState_1 = require("./useGlobalState");
var useGlobalStateReducer_1 = require("./useGlobalStateReducer");
var notImplementedErrorMsg = [
    "You must implement 'loadState' and 'saveState' to be able ",
    'to save state to your preffered storage. E.g \n',
    'store.persist({ \n',
    '    saveState: function(key, state, isInitialSet){/*logics to save state to storage*/}, \n',
    '    loadState: function(key){/*logics to load state from storage*/} \n',
    '}) \n'
].join("");
var PersistentStorage = /** @class */ (function () {
    function PersistentStorage() {
        // Persist all states in a store unless parsist = false is passed
        this.SHOULD_PERSIST_BY_DEFAULT = false;
    }
    PersistentStorage.prototype.loadState = function (key) {
        // The `if` is for tricking TS into thinking
        // this function can return some value
        if (true)
            throw TypeError(notImplementedErrorMsg);
        else
            return null;
    };
    PersistentStorage.prototype.saveState = function (key, state, isInitialSet) {
        throw TypeError(notImplementedErrorMsg);
    };
    return PersistentStorage;
}());
var Store = /** @class */ (function () {
    function Store() {
        this.value = new Map();
        this.subscriptions = [];
        this.persistentStorage = new PersistentStorage();
    }
    Store.prototype.subscribe = function (observer) {
        var _this = this;
        if (this.subscriptions.indexOf(observer) === -1) {
            // Subscribe a component to this store
            this.subscriptions.push(observer);
        }
        var unsubscribe = function () {
            _this.subscriptions = _this.subscriptions.filter(function (subscriber) { return subscriber !== observer; });
        };
        return unsubscribe;
    };
    Store.prototype.onStoreUpdate = function (key, value) {
        this.subscriptions.forEach(function (subscription) {
            subscription(key, value);
        });
    };
    Store.prototype.persist = function (config) {
        if (config.saveState) {
            PersistentStorage.prototype.saveState = config.saveState;
        }
        if (config.loadState) {
            PersistentStorage.prototype.loadState = config.loadState;
        }
        if (config.removeState) {
            PersistentStorage.prototype.removeState = config.removeState;
        }
        if (config.PERSIST_ENTIRE_STORE) {
            PersistentStorage.prototype.SHOULD_PERSIST_BY_DEFAULT = config.PERSIST_ENTIRE_STORE;
        }
    };
    Store.prototype.setState = function (key, initialValue, _a) {
        var _this = this;
        var _b = _a === void 0 ? {} : _a, persist = _b.persist;
        var shouldPersist = persist === undefined ?
            this.persistentStorage.SHOULD_PERSIST_BY_DEFAULT : persist;
        if (shouldPersist) {
            // Load state from localStorage
            var savedState = this.persistentStorage.loadState(key);
            if (savedState !== undefined) {
                // Use savedState as the initialValue
                initialValue = savedState;
            }
            else {
                // This is the initial set
                this.persistentStorage.saveState(key, initialValue, true);
            }
        }
        var onGlobalStateChange = function (newValue) {
            // Note key, persist & timerId variables depends on the scope
            _this.onStoreUpdate(key, newValue);
            if (shouldPersist) {
                _this.persistentStorage.saveState(key, newValue, false);
            }
        };
        // Create global state
        var globalState = (0, GlobalState_1.createGlobalstate)(initialValue);
        globalState.persist = shouldPersist;
        globalState.subscribe({
            observer: onGlobalStateChange,
            selector: function (state) { return state; }
        });
        // Add global state to the store
        this.value.set(key, globalState);
    };
    Store.prototype.getState = function (key, config) {
        if (config === void 0) { config = {}; }
        var defaultValue = config.default;
        // Get key based global state
        if (!this.value.has(key)) { // Global state is not found
            if (defaultValue !== undefined) { // Default value is found
                // Create a global state and use defaultValue as the initial value
                this.setState(key, defaultValue, { persist: config.persist });
            }
            else {
                // Global state is not found and the default value is not specified
                var errorMsg = [
                    "There is no global state with the key '".concat(key, "', "),
                    "You are either trying to access a global ",
                    "state which was not created or it was deleted."
                ];
                throw Error(errorMsg.join(""));
            }
        }
        return this.value.get(key);
    };
    Store.prototype.clear = function (fn) {
        var _this = this;
        // Copy store
        var storeCopy = this.value;
        // Clear store
        this.value = new Map();
        if (this.persistentStorage.clearStorage) {
            this.persistentStorage.clearStorage();
        }
        if (fn) {
            // Run store re-initialization
            fn();
        }
        storeCopy.forEach(function (oldState, key) {
            // Notify subscribers to a store that a global state has been removed
            if (_this.value.has(key)) {
                var newGlobalState = _this.getState(key);
                _this.onStoreUpdate(key, newGlobalState.getValue());
            }
            // Rerender all components using this global state
            oldState.refresh();
        });
    };
    Store.prototype.remove = function (globalStatekey, fn) {
        var _this = this;
        var keys = [];
        if (typeof globalStatekey === 'string') {
            keys = [globalStatekey];
        }
        else {
            keys = globalStatekey;
        }
        var globalStatesToRemove = new Map();
        keys.forEach(function (key) {
            // Copy global state to remove from a store
            globalStatesToRemove.set(key, _this.getState(key));
            // Remove global state from a store
            _this.value.delete(key);
            if (_this.persistentStorage.removeState && globalStatesToRemove.get(key).persist) {
                _this.persistentStorage.removeState(key);
            }
        });
        if (fn) {
            // Run global state re-initialization
            fn();
        }
        globalStatesToRemove.forEach(function (oldState, key) {
            // Notify subscribers to a store that a global state has been removed
            if (_this.value.has(key)) {
                var newGlobalState = _this.getState(key);
                _this.onStoreUpdate(key, newGlobalState.getValue());
            }
            // Rerender all components depending on this global state
            oldState.refresh();
        });
    };
    Store.prototype.useState = function (key, config) {
        if (config === void 0) { config = {}; }
        var globalState = this.getState(key, config);
        return (0, useGlobalState_1.useGlobalState)(globalState, config);
    };
    Store.prototype.useReducer = function (reducer, key, config) {
        if (config === void 0) { config = {}; }
        var globalState = this.getState(key, config);
        return (0, useGlobalStateReducer_1.useGlobalStateReducer)(reducer, globalState, config);
    };
    return Store;
}());
exports.Store = Store;
function createStore() {
    // Create store for key based global state
    return new Store();
}
exports.createStore = createStore;
