// state-pool/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useGlobalState = exports.useGlobalStateReducer = exports.createStore = exports.Store = exports.createGlobalstate = exports.GlobalState = void 0;
var GlobalState_1 = require("./GlobalState");
Object.defineProperty(exports, "GlobalState", { enumerable: true, get: function () { return GlobalState_1.GlobalState; } });
Object.defineProperty(exports, "createGlobalstate", { enumerable: true, get: function () { return GlobalState_1.createGlobalstate; } });
var Store_1 = require("./Store");
Object.defineProperty(exports, "Store", { enumerable: true, get: function () { return Store_1.Store; } });
Object.defineProperty(exports, "createStore", { enumerable: true, get: function () { return Store_1.createStore; } });
var useGlobalStateReducer_1 = require("./useGlobalStateReducer");
Object.defineProperty(exports, "useGlobalStateReducer", { enumerable: true, get: function () { return useGlobalStateReducer_1.useGlobalStateReducer; } });
var useGlobalState_1 = require("./useGlobalState");
Object.defineProperty(exports, "useGlobalState", { enumerable: true, get: function () { return useGlobalState_1.useGlobalState; } });
